File: scripts/forest_plot.py
#!/usr/bin/env python3
"""
Forest Plot Generator
For displaying multivariable regression results:
- Variable names, effect size (OR/HR/beta), 95% CI, P-value
- Exports to PNG + PDF
Medical Statistics Skill
"""
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from datetime import datetime


def forest_plot(data, effect_label="OR (95% CI)", reference=1,
                filename_prefix="forest", title=None, footnote=None):
    """
    Generate a publication-quality forest plot.

    Parameters
    ----------
    data : pd.DataFrame
        Must contain columns: Variable, Effect, Lower, Upper, [P_value]
    effect_label : str
        Label for x-axis ("OR (95% CI)", "HR (95% CI)", "beta (95% CI)")
    reference : float
        Reference value (1 for OR/HR, 0 for beta)
    filename_prefix : str
        Output file prefix
    title : str
        Plot title
    footnote : str
        Caption text
    """
    required = ["Variable", "Effect", "Lower", "Upper"]
    missing = [c for c in required if c not in data.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")

    df = data.copy()

    # Significance stars
    if "P_value" in df.columns:
        def star(p):
            if pd.isna(p):
                return ""
            elif p < 0.001:
                return "***"
            elif p < 0.01:
                return "**"
            elif p < 0.05:
                return "*"
            else:
                return "ns"
        df["Sig"] = df["P_value"].apply(star)
    else:
        df["Sig"] = ""

    # Display text
    if "P_value" in df.columns:
        df["DisplayText"] = df.apply(
            lambda r: f"{r['Effect']:.3f} ({r['Lower']:.3f}–{r['Upper']:.3f})  {r['P_value']:.4f}{r['Sig']}",
            axis=1
        )
    else:
        df["DisplayText"] = df.apply(
            lambda r: f"{r['Effect']:.3f} ({r['Lower']:.3f}–{r['Upper']:.3f})",
            axis=1
        )

    # Reverse order so first row is at top
    df = df.iloc[::-1].reset_index(drop=True)
    df["y_pos"] = range(len(df))

    # Set up the plot
    use_log = reference == 1
    n_vars = len(df)

    fig_height = max(4, n_vars * 0.5 + 1.5)
    fig, ax = plt.subplots(figsize=(10, fig_height))

    # X axis range
    all_vals = pd.concat([df["Lower"], df["Upper"], pd.Series([reference])])
    x_min, x_max = all_vals.min(), all_vals.max()
    x_pad = (x_max - x_min) * 0.15 if x_max != x_min else 0.5
    x_lim = (x_min - x_pad, x_max + x_pad)

    # Reference line
    ax.axvline(x=reference, color="#b2182b", linestyle="--", linewidth=0.8, alpha=0.7, zorder=1)

    # CI lines
    for _, row in df.iterrows():
        ax.plot([row["Lower"], row["Upper"]], [row["y_pos"], row["y_pos"]],
                color="gray30", linewidth=1.5, solid_capstyle="round", zorder=2)

    # Point estimates with color by significance
    color_map = {"***": "#b2182b", "**": "#d6604d", "*": "#f4a582", "ns": "gray60", "": "gray40"}
    for _, row in df.iterrows():
        color = color_map.get(row["Sig"], "gray40")
        ax.scatter(row["Effect"], row["y_pos"], color=color, s=60, zorder=3, edgecolors="white", linewidth=0.5)

    # Text labels on the right
    for _, row in df.iterrows():
        ax.text(x_lim[1], row["y_pos"], row["DisplayText"],
                ha="right", va="center", fontsize=8.5, fontfamily="serif")

    # Y axis labels (variable names)
    ax.set_yticks(df["y_pos"])
    ax.set_yticklabels(df["Variable"], fontsize=10, fontfamily="serif")

    # X axis
    if use_log:
        ax.set_xscale("log")
        ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f"{x:.2f}"))

    ax.set_xlabel(effect_label, fontsize=11, fontweight="bold")
    ax.set_title(title or f"Forest Plot: {effect_label}",
                 fontsize=13, fontweight="bold", pad=12)

    # Caption / footnote
    caption = footnote or "Significance: *** P<0.001, ** P<0.01, * P<0.05, ns = not significant"
    ax.set_xlim(x_lim)
    ax.grid(True, axis="x", alpha=0.25, linestyle=":")
    ax.grid(False, axis="y")
    ax.axhline(y=-0.5, color="gray80", linewidth=0.5)

    # Reference annotation
    if use_log:
        ax.text(reference, ax.get_ylim()[0] - 0.3, "Reference",
                ha="center", fontsize=8, color="#b2182b", alpha=0.6)

    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.spines["left"].set_color("gray80")
    ax.spines["bottom"].set_color("gray80")

    fig.text(0.02, 0.01, caption, fontsize=7.5, color="gray50", fontfamily="serif")

    plt.tight_layout()

    # Export
    print(f"\n  Generating forest plot: {title or filename_prefix}")

    png_file = f"{filename_prefix}.png"
    fig.savefig(png_file, dpi=300, bbox_inches="tight")
    print(f"  ✔ PNG: {png_file}")

    pdf_file = f"{filename_prefix}.pdf"
    fig.savefig(pdf_file, format="pdf", bbox_inches="tight")
    print(f"  ✔ PDF: {pdf_file}")

    plt.close(fig)
    return fig


def forest_plot_from_model(result_df, model_type, ts=None):
    """Convenience wrapper: create forest plot from model results DataFrame."""
    if ts is None:
        ts = datetime.now().strftime("%Y%m%d_%H%M%S")

    if model_type == "linear":
        effect_label = "beta (95% CI)"
        reference = 0
    elif model_type == "cox":
        effect_label = "HR (95% CI)"
        reference = 1
    else:
        effect_label = "OR (95% CI)"
        reference = 1

    titles = {"linear": "Linear Regression", "logistic": "Logistic Regression",
              "cox": "Cox Proportional Hazards"}
    prefix = f"forest_{model_type}_{ts}"

    forest_plot(result_df, effect_label=effect_label, reference=reference,
                filename_prefix=prefix, title=f"Forest Plot: {titles.get(model_type, model_type)}")
    return prefix


if __name__ == "__main__":
    print("✔ forest_plot.py loaded")

File: scripts/forest_plot.R
#!/usr/bin/env Rscript
# ============================================
# 森林图 (Forest Plot) — 使用 forestplot 包
# 参考: 腾讯云开发者社区 forestplot 教程
# 左侧: 变量名 | 系数 | P值 | OR/HR (95% CI)
# 右侧: 森林图 (方块 + CI线)
# 导出 PNG + PDF
# medical-statistics skill
# ============================================

#' 生成发表级森林图
#'
#' @param data 数据框，必须含: Variable, Effect, Lower, Upper, P_value
#' @param effect_label 效应标签 ("OR (95% CI)" / "HR (95% CI)" / "beta (95% CI)")
#' @param reference 参考值 (OR/HR=1, beta=0)
#' @param filename_prefix 输出前缀
#' @param title 标题
#' @param footnote 脚注
#'
forest_plot <- function(data, effect_label = "OR (95% CI)", reference = 1,
                        filename_prefix = "forest", title = NULL, footnote = NULL) {

  # ── 加载/安装 forestplot ──
  if (!requireNamespace("forestplot", quietly = TRUE)) {
    install.packages("forestplot", repos = "https://cran.r-project.org")
  }
  library(forestplot)
  library(grid)

  # 参数校验
  required_cols <- c("Variable", "Effect", "Lower", "Upper")
  missing <- setdiff(required_cols, names(data))
  if (length(missing) > 0) {
    stop("数据缺少必需列: ", paste(missing, collapse = ", "))
  }

  df <- data
  n <- nrow(df)

  # ── 准备显示文本 ──
  # 效应值文本: OR/HR = 1.234 (1.012-1.456)
  eff_text <- sprintf("%.3f (%.3f–%.3f)", df$Effect, df$Lower, df$Upper)

  # P值文本
  p_text <- ifelse(df$P_value < 0.001, "<0.001",
                   sprintf("%.3f", round(df$P_value, 3)))

  # 系数文本（OR/HR 用 log 尺度显示系数）
  if (reference == 1) {
    coef_text <- sprintf("%.3f", log(df$Effect))
  } else {
    coef_text <- sprintf("%.3f", df$Effect)
  }

  # ── 表头 ──
  # 根据效应类型设置表头
  if (reference == 1) {
    header_coef <- "log(OR/HR)"
    header_eff  <- effect_label
  } else {
    header_coef <- "Coefficient"
    header_eff  <- "beta (95% CI)"
  }

  # 构建 labeltext 矩阵（四列: 变量名, 系数, P值, 效应值）
  tabletext <- cbind(
    c(NA, "Variable", df$Variable),
    c(NA, header_coef, coef_text),
    c(NA, "P value", p_text),
    c(NA, header_eff, eff_text)
  )

  # ── 颜色 ──
  # 参考文章: box = "#D55E00", lines = "#CC79A7", zero = "gray50"
  fp_col <- fpColors(
    box    = "#2166ac",
    lines  = "#4393c3",
    zero   = "gray50",
    hrz_lines = "#333333"
  )

  # ── 确定图形位置和 X 轴范围 ──
  if (reference == 1) {
    # OR/HR: 对数尺度
    x_ticks <- pretty(c(df$Lower, df$Upper, reference), n = 5)
    x_ticks <- x_ticks[x_ticks > 0]
  } else {
    x_ticks <- pretty(c(df$Lower, df$Upper, reference), n = 5)
  }

  # ── 水平分割线 ──
  hrzl <- list("2" = gpar(lwd = 1.5, col = "#333333"))
  # 最后一行下方
  hrzl[[as.character(n + 2)]] <- gpar(lwd = 1.5, col = "#333333")

  # ── 绘制森林图 ──
  p <- forestplot(
    labeltext  = tabletext,
    mean       = c(NA, NA, df$Effect),
    lower      = c(NA, NA, df$Lower),
    upper      = c(NA, NA, df$Upper),
    is.summary = c(TRUE, TRUE, rep(FALSE, n)),

    # 图形位置（第4列之后）
    graph.pos = 4,

    # 颜色
    col = fp_col,

    # X轴
    zero    = reference,
    xlab    = effect_label,
    xticks  = x_ticks,
    xlog    = (reference == 1),

    # CI 线样式
    boxsize       = 0.15,
    lwd.ci        = 1.5,
    ci.vertices   = TRUE,
    ci.vertices.height = 0.08,

    # 行高和图宽
    lineheight = unit(0.8, "cm"),
    graphwidth = unit(0.35, "npc"),
    colgap     = unit(4, "mm"),

    # 字体
    txt_gp = fpTxtGp(
      label   = gpar(fontfamily = "serif", cex = 0.85),
      ticks   = gpar(fontfamily = "serif", cex = 0.8),
      xlab    = gpar(fontfamily = "serif", cex = 0.9, fontface = "bold"),
      title   = gpar(fontfamily = "serif", cex = 1, fontface = "bold")
    ),

    # 分割线
    hrzl_lines = hrzl,

    # 标题
    title = if (!is.null(title)) title else "",

    # 边距
    mar = unit(c(0, 0, 0, 0), "mm"),

    # 不显示 ICE (influence curve)
    fn.ci_norm = fpDrawCircleCI,

    # 抑制冗余输出
    new_page = TRUE
  )

  # ── 导出 ──
  cat(sprintf("\n  📊 生成森林图: %s\n",
              if (!is.null(title)) title else filename_prefix))

  # PNG: 使用 grDevices 的 png() 设备
  png_file <- paste0(filename_prefix, ".png")
  png(png_file, width = 3000, height = max(1200, n * 200 + 600), res = 300)
  attr(p, "forestplot") <- NULL  # 防止双重复制

  # 重新绘制
  forestplot(
    labeltext  = tabletext,
    mean       = c(NA, NA, df$Effect),
    lower      = c(NA, NA, df$Lower),
    upper      = c(NA, NA, df$Upper),
    is.summary = c(TRUE, TRUE, rep(FALSE, n)),
    graph.pos  = 4,
    col        = fp_col,
    zero       = reference,
    xlab       = effect_label,
    xticks     = x_ticks,
    xlog       = (reference == 1),
    boxsize    = 0.15,
    lwd.ci     = 1.5,
    ci.vertices = TRUE,
    ci.vertices.height = 0.08,
    lineheight = unit(0.8, "cm"),
    graphwidth = unit(0.35, "npc"),
    colgap     = unit(4, "mm"),
    txt_gp = fpTxtGp(
      label   = gpar(fontfamily = "serif", cex = 0.85),
      ticks   = gpar(fontfamily = "serif", cex = 0.8),
      xlab    = gpar(fontfamily = "serif", cex = 0.9, fontface = "bold"),
      title   = gpar(fontfamily = "serif", cex = 1, fontface = "bold")
    ),
    hrzl_lines = hrzl,
    title = if (!is.null(title)) title else "",
    mar = unit(c(0, 0, 0, 0), "mm"),
    fn.ci_norm = fpDrawCircleCI,
    new_page = TRUE
  )
  invisible(dev.off())
  cat(sprintf("  ✔ PNG: %s\n", normalizePath(png_file)))

  # PDF
  pdf_file <- paste0(filename_prefix, ".pdf")
  pdf(pdf_file, width = 12, height = max(4, n * 0.65 + 2))
  # 重新绘制
  forestplot(
    labeltext  = tabletext,
    mean       = c(NA, NA, df$Effect),
    lower      = c(NA, NA, df$Lower),
    upper      = c(NA, NA, df$Upper),
    is.summary = c(TRUE, TRUE, rep(FALSE, n)),
    graph.pos  = 4,
    col        = fp_col,
    zero       = reference,
    xlab       = effect_label,
    xticks     = x_ticks,
    xlog       = (reference == 1),
    boxsize    = 0.15,
    lwd.ci     = 1.5,
    ci.vertices = TRUE,
    ci.vertices.height = 0.08,
    lineheight = unit(0.8, "cm"),
    graphwidth = unit(0.35, "npc"),
    colgap     = unit(4, "mm"),
    txt_gp = fpTxtGp(
      label   = gpar(fontfamily = "serif", cex = 0.85),
      ticks   = gpar(fontfamily = "serif", cex = 0.8),
      xlab    = gpar(fontfamily = "serif", cex = 0.9, fontface = "bold"),
      title   = gpar(fontfamily = "serif", cex = 1, fontface = "bold")
    ),
    hrzl_lines = hrzl,
    title = if (!is.null(title)) title else "",
    mar = unit(c(0, 0, 0, 0), "mm"),
    fn.ci_norm = fpDrawCircleCI,
    new_page = TRUE
  )
  invisible(dev.off())
  cat(sprintf("  ✔ PDF: %s\n", normalizePath(pdf_file)))

  # PNG 副本（低分辨率预览）
  return(invisible(p))
}

#' 从多因素模型结果生成森林图（便利函数）
forest_plot_from_model <- function(result_df, model_type, ts = NULL) {
  if (is.null(ts)) ts <- format(Sys.time(), "%Y%m%d_%H%M%S")

  effect_label <- switch(model_type,
    "linear"   = "beta (95% CI)",
    "cox"      = "HR (95% CI)",
    "logistic" = "OR (95% CI)",
    "OR (95% CI)"
  )
  reference <- if (model_type == "linear") 0 else 1
  title_map <- c(linear = "Linear Regression", logistic = "Logistic Regression",
                 cox = "Cox Proportional Hazards")

  prefix <- paste0("forest_", model_type, "_", ts)
  forest_plot(result_df,
              effect_label = effect_label,
              reference    = reference,
              filename_prefix = prefix,
              title = sprintf("Forest Plot: %s", title_map[model_type]))
  return(prefix)
}

if (sys.nframe() == 0) {
  cat("✔ forest_plot.R loaded (forestplot package)\n")
}
